src/main/java/org/apache/maven/plugins/assembly/artifact/ScopeBasedDependencySelector.java

```diff
@@ -0,0 +1,112 @@
+package org.apache.maven.plugins.assembly.artifact;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.maven.artifact.Artifact;
+import org.sonatype.aether.collection.DependencyCollectionContext;
+import org.sonatype.aether.collection.DependencySelector;
+import org.sonatype.aether.graph.Dependency;
+
+/**
+ * Dependency selector which based on the hierarchy of scopes, calculates which dependency should be included or not.
+ *
+ */
+public class ScopeBasedDependencySelector implements DependencySelector
+{
+
+    private boolean compileScope;
+
+    private boolean runtimeScope;
+
+    private boolean testScope;
+
+    private boolean providedScope;
+
+    private boolean systemScope;
+
+    public ScopeBasedDependencySelector( String scope )
+    {
+        switch ( scope )
+        {
+            case Artifact.SCOPE_PROVIDED:
```

Comment:
This looks like there's a missing type somewhere, perhaps an enum, which would have these values as properties